### ISSUE TITLE
add Hampel filter

### ISSUE THREAD

**Issue by jerabaul29** [scipy/scipy#12809]
I would like to perform outliers detection in some time series, a quite common task in many workflows.

One simple method is the Hampel filtering: cut the data in segments, then for each segment, throw away data that deviate more than a given number of times the std of the segment.

## context and quick internet search

This is implemented in other "concurrents" to python, or in some other python packages:

https://www.rdocumentation.org/packages/pracma/versions/1.9.9/topics/hampel

https://se.mathworks.com/help/dsp/ref/hampelfilter.html

https://se.mathworks.com/help/dsp/ref/dsp.hampelfilter-system-object.html

https://se.mathworks.com/help/signal/ref/hampel.html

https://python-heart-rate-analysis-toolkit.readthedocs.io/en/latest/heartpy.filtering.html

## points suggesting this should be implemented

- I could not find this filter in scipy, do you agree that this is not available at the moment?
- This is a simple, robust, "default go to" method for outlier detection in time series; do you agree?

If both points are true, should it be implemented?

## Implementation

I may be interested in helping implement (no guarantee though and if some more experienced scipy dev can do it, it would probably be better), but I am not too used with how Scipy works. What would be helpful to you? A python implementation? A c++ implementation for speed that you may wrap through python and numpy arrays?

**Comment by jerabaul29** [comment]
I can confirm that the pure python + numpy implementation is veeeery slow.

**Comment by ilayn** [comment]
> -  I could not find this filter in scipy, do you agree that this is not available at the moment?
> - This is a simple, robust, "default go to" method for outlier detection in time series; do you agree?
> If both points are true, should it be implemented?

For what its worth, my response is Yes * 3. A cythonized version would be sufficiently quick I would feel (with no basis for my claim). 

@larsoner @ev-br @endolith and others; thoughts?

**Comment by jerabaul29** [comment]
That sounds great :) 

A few comments / brainstorming after playing a bit around with this filter in the last days:

- I think that it would be great to use the filter only for returning an array of points to remove, or the array with the problematic points substituted by NaNs, i.e. that the function does not perform any interpolation or else, so that the user can do whatever he wants after the filter has been called.

- I wonder if there are several possible variations around this filter. I was a bit surprised that it relies on the median, and not on the mean.

- I also wonder if an alternative to this filter could be some form of Chauvenet criterion and if one could implement it also (they are relatively similar, so when people get into the technical details of making sure things are implemented right, may be a good idea to implement both while ideas are fresh around those criteria :) ).

**Comment by ev-br** [comment]
Not sure I can bring anything to this one other than saying feel free to ping me if you need a hand cythonizing.

**Comment by jerabaul29** [comment]
What about discussing here the function then? Or do you want to discuss through some pull request of something like this?

Currently I am using the following on my small project. Note that I am not specialized in statistics, this is taken from the SO answer and a bit modified to fit my need.

```python
import numpy as np


def hampel(x, k, t0=3, exclude_crrt_point=True, nan_substitution=True):
    '''Perform hampel filtering, returning both filtered series and
    mask of filtered points.
    Input:
        x: 1-d numpy array of numbers to be filtered
        k: number of items in (window-1)/2, i.e. using k elems before and after
            in addition to the current point
        t0: number of standard deviations to use; 3 is default
        exclude_crrt_point: if the current point being inspected should be
            ignored when calculating variance.
        nan_substitution: if True then invalid points should be replaced by a
            NaN value. If False, interpolate with the median value.
    Output:
        y: 1-d numpy array obtained by filtering (this is a np.copy of x),
            NaN where discarded
        mask_modified: boolean mask, True if point has been discarded or
            modified by the filter
    '''

    # NOTE: adapted from hampel function in R package pracma
    # NOTE: this is adapted from: https://stackoverflow.com/questions/46819260/
    # filtering-outliers-how-to-make-median-based-hampel-function-faster
    # NOTE: adapted from the issue by Jean Rabault jean.rblt@gmail.com 09-2020

    if not isinstance(x, np.ndarray):
        raise ValueError("x should be a numpy array")

    if not len(np.squeeze(x).shape) == 1:
        raise ValueError("x should be 1-dimensional")

    if not isinstance(k, int):
        raise ValueError("k should be an int")

    if not isinstance(t0, int):
        raise ValueError("t0 should be an int")

    y = np.copy(x)  # y is the corrected series
    y = np.squeeze(y)

    mask_modified = np.zeros((y.shape[0]), dtype=bool)

    n = y.shape[0]
    L = 1.4826

    # cannot apply the filter too close to the edges, as we need k points
    # before and after
    for i in range((k), (n - k)):

        # follow user preference on using or not the current point for
        # estimating statistical properties
        if exclude_crrt_point:
            array_neighborhood = np.concatenate((
                y[(i - k):(i)],
                y[(i + 1):(i + k + 1)]
            ))
        else:
            array_neighborhood = y[i - k: i + k + 1]

        # if all points around are already nans, cannot trust local point
        if np.isnan(array_neighborhood).all():
            if not np.isnan(y[i]):
                y[i] = np.nan
                mask_modified[i] = True
            continue

        # if current point is already a nan, keep it so
        if np.isnan(y[i]):
            continue

        # otherwise, should perform the filtering
        x0 = np.nanmedian(array_neighborhood)
        S0 = L * np.nanmedian(np.abs(array_neighborhood - x0))

        if (np.abs(y[i] - x0) > t0 * S0):
            if nan_substitution:
                y[i] = np.nan
            else:
                y[i] = x0
            mask_modified[i] = True

    return (y, mask_modified)


if __name__ == "__main__":
    # test 1
    x = np.arange(0, 2*np.pi, 2*np.pi/100)
    y = np.sin(x)
    y[15] = 120
    y[30] = np.nan

    filtered, mask = hampel(y, 5)
    print(filtered)
    assert np.isnan(filtered[15])
    assert (np.where(np.isnan(filtered))[0] == np.array([15, 30])).all()


```

**Comment by jerabaul29** [comment]
@ev-br this is quite slow, so if you want to cythonize that would be great :) .

**Comment by jerabaul29** [comment]
Any comments / how do you want to proceed further on this? :)

**Comment by adrianomitre** [comment]
In the paper [On computation of the running median](https://ieeexplore.ieee.org/document/17539), authored  by J.T. Astola and T.G. Campbell, it is presented an algorithm with θ(_n_ ⋅ log(_m_)) time complexity, in constrast with θ(_n_ ⋅ _m_ log(_m_)) of the straightforward approach, where _n_ is the signal length and _m_ is the window length.

The central idea is employing a min-heap and a max-heap, which sharing a single element (the median):
<p align="center">
<img src="https://user-images.githubusercontent.com/456804/124969437-13fd0100-dffd-11eb-8342-cd5c998f3bb5.png" width="320">
</p>

**Comment by adrianomitre** [comment]
Another interesting, more recent, reference is [Median Filtering is Equivalent to Sorting](https://arxiv.org/abs/1406.1717) authored by Jukka Suomela. At the end of the paper there is a Python code listing. Also, there is an implementation by the author [here](https://github.com/suomela/median-filter), including the [Python implementation listed in the paper](https://raw.githubusercontent.com/suomela/median-filter/master/python/SortMedian.py), in addition to a C++11 implementation and reference to a 2D median filter from the same author.

**Comment by mretegan** [comment]
@jerabaul29 The time-consuming part of the Hampel filter is the computation of the rolling window median. Here is a link to a notebook comparing your function and another one that uses views and eliminates the for loop (https://gist.github.com/mretegan/5ec039a216449f5a9bbf81854f7919a2). It is about 40 times faster.

**Comment by adrianomitre** [comment]
The implementation [in this Gist](https://gist.github.com/adrianomitre/2ced5544421eca031180782cc30798af), is more than 6x faster than the fastest implementation on the notebook, which uses sliding tricks, and more than 220x faster than the slower one. I must point out that these comparison do not account for the ~1 second it takes to JIT compile in the very first run.  Also, it depends on [numba](http://numba.pydata.org/).

**Comment by ilayn** [comment]
Another issue that I criminally forgot to attend. My apologies @jerabaul29 . And thanks for the discussion and references @adrianomitre and @mretegan. 

If we can stabilize over the method to go then it seems to me that sorting and other loops are well suited for Cythonization. It seems to me that @suomela is using lists as poor man's linked list replacement but we can handle that a bit better. Hence performance can be increased a lot compared to direct Python code. If folks are still not satisfied we can also look into the binding the C++ version though that makes the maintenance trickier.

**Comment by roger-** [comment]
Would love to see this get added at some point. What are the next steps?

**Comment by mckib2** [comment]
> Another interesting, more recent, reference is [Median Filtering is Equivalent to Sorting](https://arxiv.org/abs/1406.1717) authored by Jukka Suomela. At the end of the paper there is a Python code listing. Also, there is an implementation by the author [here](https://github.com/suomela/median-filter), including the [Python implementation listed in the paper](https://raw.githubusercontent.com/suomela/median-filter/master/python/SortMedian.py), in addition to a C++11 implementation and reference to a 2D median filter from the same author.

Just checked, the README does say the code in that repo can be distributed/used under the MIT license, so we can reference and use it.  Might want to contact the repo owner there to suggest they add the license to the project instead of just a note in the README

**Comment by tstorek** [comment]
Hello everybody,
is there any news on this? I currently, want to use a Hampel filter for a time series analysis project. It would be great to simply use the implementation from scipy instead of an additional package or implementing it by myself. 
Especially, since this issue is very close to the end, right?

**Comment by gideonKogan** [comment]
I like to peak up the task. I am familiar with double heap and median filter. IMO, this problem is slightly different as after the median we have to implement another median for the "median absolute deviation". IMO, a good open-source reference is given [here](https://www.gnu.org/software/gsl/doc/html/filter.html?highlight=hampel#impulse-detection-filter). Unfortunatly, I am afraid to open it as I am not sure we can reuse the implementation afterward. Are we allowed to use it or read it before the implementation? I am not sufficiently familiar with all the legal aspects...

**Comment by gideonKogan** [comment]
ok' I have an idea for an implementation. It should maintain the same complexity order as the median filter. Adding another double heap with the pointers for the values maintaining the median absolute deviation (MAD). Here, I avoid creating a double heap for the MAD itself as I will have to update it every time that the median is updated but having only the pointers to the values should be ok as the median is pretty stable and a new value should be not more than one element down/up the heap. It might take me a couple of days to get it done though (other obligations). 
Please tell me what you think of it as did not find a reference in the literature and I am still afraid to look into the code above...
PS: I have an old Python implementation of mine of the median filter based on 2011 ashelly.myopenid.com under **mit-license**. I have slightly modified it to reduce the computation. Now it requires some initiation but it should run faster and the code is simpler. Can I use it? I can create a new implementation too...

**Comment by lucascolley** [comment]
MIT licensed code is fine to include 👍 

cc @DietBru as you've been active in Signal recently!

**Comment by josef-pkt** [comment]
both scipy signal and ndimage already have median filter

I think MAD filtering can be done with 2 medfilt plus vectorized (full array) computations.
(I might even have done something like that in the past)

**Comment by josef-pkt** [comment]
I also just saw https://github.com/MichaelisTrofficus/hampel_filter
with basic hampel filter in cython

**Comment by DietBru** [comment]
I cannot contribute too much to this topic: 
* I am bit a uncomfortable with the fiddle factor of  `L = 1.4826`, which supposedly relates the median to the standard deviation. This [only holds](https://en.wikipedia.org/wiki/Median_absolute_deviation#Relation_to_standard_deviation) if the signals samples are [normally distributed](https://en.wikipedia.org/wiki/Normal_distribution). Looking at the examples of the provided links, I think it is safe to say that not [Gaussian noise](https://en.wikipedia.org/wiki/Gaussian_noise)  but  deterministic components dominate a typical signal. Hence, not the number of standard deviations `t_0`, but a fiddle factor $\xi:=t_0 L$ should be a parameter for [hempel()](https://github.com/scipy/scipy/issues/12809#issuecomment-689505404). The factor of `L = 1.4826` could be explained in the docstring.
* For data with a constant sampling interval, doing appropriate high-pass filtering (think AC-coupling) and using $k\times$ the standard deviation (i.e., the r.m.s. value) of a sliding window as a threshold is probably a more suitable heuristic than MAD. The reason being that such a signal typically represents a band-limited continuous-time signal, where a median amplitude is a bit tedious to define and not straight-forward to reason about.
* Irregularly sampled data with single sample outliers is the scenario where the Hampel filter shines, IMHO. So it might be beneficial to put such an example into the documentation.
* It would be nice if the documentation could address some of the edge case, like the ones @josef-pkt provided. It would be even better, if it would be possible to convey the point that the quality of the outlier detection depends on how well the signal matches the statistical assumptions the detector makes on how the signal "should" be behave. 
* For consistency reasons, it might make sense to
   * allow complex-valued signals due to #19771 and
   * have an `axis` keyword so multi-dimensional arrays can be filtered as well. That would provide the same behavior in that regard as [fft()](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fft.html) or  [filtfilt()](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.filtfilt.html).

**Comment by josef-pkt** [comment]
> Hence, not the number of standard deviations t_0, but a fiddle factor 
 should be a parameter for [hempel()](https://github.com/scipy/scipy/issues/12809#issuecomment-689505404).

I think current usage is easier to interpret

both t_0 and scaling in MAD to get S0 change with the reference distribution

t_0 is the critical value coming from the distributional assumption and tail probability e.g. norm.ppf(0.99). If we have a fat tailed distribution, then we should go for a larger sigma factor, 3sigma, 5sigma, ...
S0, scaled mad is just an estimate of standard deviation, other robust scale estimates are standardized in the same way. It is difficult to compute or figure out the correct distribution dependent scaling factor of those estimators for other distributions.

We need only one parameter to affect outlier threshold and I think t_0 is easier to interpret. 
Recomputing the correct threshold would for a specific non-normal distribution would require to recompute both factors, but I don't anyone will do that.

**Comment by DietBru** [comment]
@josef-pkt, I fully agree with you that for signals, where the dominant parts can be modeled by a Gaussian process (including the outliers!), the $\pm t_0 \sigma$-band is the natural way to distinguish outliers from valid values.
I am questioning if such distributional assumptions are relevant for real-world problems requiring an outlier detection. For such stochastic processes, Wiener or Kalman Filters seem a more fitting tool to me. 
What would be a practical example where it is reasonable assume that the samples of the sliding window used to calculate the median are somewhat Gaussian?

**Comment by gideonKogan** [comment]
As has been mentioned, the compromise is always on the scale of robustness vs efficiency. All the cases for which robust statistics useful, apply here too. A minor example is small drift errors that accumulate into overflow spikes, or partial discharges. 

I have implemented in **Python** a [solution](https://github.com/ggkogan/hampel/tree/main) with a logaritmic complexity order. Obviously, the linear solution should be more efficient for smaller windows and the attached solution will work efficiently for larger windows. I assume that after some optimization, we can define the spot of transition from the linear solution to this one. I would be happy to know what you think of if and if it is a good base to use as a fundamental for the complete solution. In case it is, I would be happy to proceed to Cython/C implementation with an integration of one of the linear solutions into it.

**Comment by josef-pkt** [comment]
I'm fighting with similar scaling issues in statsmodels https://github.com/statsmodels/statsmodels/pull/8129
Essentially working with normal distribution as reference works. If distributions are non-normal than getting the scale correctly calibrated would require knowing the distribution to compute the correct adjustment factor.
AFAICS, the is no robust estimator for the standard deviation that can be calibrated independent of the distribution.

However for huge outliers as in the examples here it does not really matter.
i.e. we can detect outliers that are e.g. 5 * calibrated mad or 10 * times calibrated mad.
And the proposed algorithm is relatively straightforward for this. (aside from performance improvements).

To detect outliers that are close to reasonably likely observations is impossible without knowing or estimating the error distribution.
(related: control charts can be make nonparametric if we can estimate quantiles of the error distribution from a clean sample.
But that gets into much more elaborate algorithms which also might require other assumptions on the data process.)

I guess I'm +1, if I had an opinion of what should go into scipy.signal.

**Comment by DietBru** [comment]
After pondering some more: The [filtering functions](https://docs.scipy.org/doc/scipy/reference/signal.html#filtering) in `scipy.signal` all demand equally spaced samples whereas the Hampel filter (as typical for method of time series analysis) does not have that limitation. As noted [above](https://github.com/scipy/scipy/issues/12809#issuecomment-1894627210), for signals with constant sampling rate I think a Hampel filter is not the best fit for many applications.
Furthermore `scipy.signal`'s filters transform the value of each sample (more or less) whereas  in outlier detection  a subset of samples is marked. Hence, the term "filtering" is used to represent two distinct concepts. For that I reason I suggest:
1. Place the Hampel filter into the [peak finding](https://docs.scipy.org/doc/scipy/reference/signal.html#peak-finding) group and rename that group to something akin to "Peak finding and Outlier detection".
2. A knit-pick: Perhaps naming the function something like `find_outliers_mad()` would be more descriptive in the context of `scipy.signal`.

In a more general context the questions arises if the Hampel filter is still in the scope of `scipy.signal`? Almost all existing  functions adhere to the philosophy of classical digital signal processing, i.e., requiring constant sample rate signals (the [peak finding](https://docs.scipy.org/doc/scipy/reference/signal.html#peak-finding) functions are the prominent exception). In my opinion, the Hampel filter, like the Kalman filter, belong more to the field of time series analysis and statistics. In summary: Is it part of `scipy.signal`'s roadmap of incorporating more non-classical statistical methods?

IMHO, a few additional functions for outlier detection would not hurt. Hence, I'm +½, , if I had an opinion of what should go into scipy.signal.

**Comment by gideonKogan** [comment]
@DietBru , would you suggest putting it next to the median filter or the order filter? I assume that both do not require a constant sampling rate too...

**Comment by DietBru** [comment]
Fair point @gideonKogan, I overlooked those. I would put them in the [Filters based on order statistics](https://docs.scipy.org/doc/scipy/tutorial/ndimage.html#filters-based-on-order-statistics) section of [scipy.ndimage](https://docs.scipy.org/doc/scipy/reference/ndimage.html). After looking through that package, I think it could be argued that the Hampel filter might there be in better company as well. 

Though the Hampel filter divides the samples into two partitions and does not modify their values. Hence it is conceptually still more akin to the [find_peaks()](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html) function than to [order_filter()](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.order_filter.html).

**Comment by ggkogan** [comment]
[attaching the previously discussed code](https://github.com/ggkogan/hampel/blob/main/hampel.cpp). I played with it. [The performance seems to be good](https://github.com/ggkogan/hampel/blob/main/res_review.ipynb). please review the notebook and tell me what you think of it. I plan starting the integration after [the 1d order filter PR](https://github.com/scipy/scipy/pull/20543) will be merged.
